File: cplus/leetcode/202511/algo0777.cpp
#include <bits/stdc++.h>
using namespace std;

// 滑动窗口与双指针 - 双序列双指针 - 4.1 双指针
class Solution {
public:
    bool canTransform(string start, string result) {
        // 从 python 翻译过来
        int n = start.length();
        int i = 0 , j = 0;
        while (i < n  || j < n) {
            while (i < n && start[i] == 'X') i ++;
            while (j < n && result[j] == 'X') j ++;

            if (i == n || j == n) break;
            
            if (start[i]==result[j] && ((start[i] == 'L' && i >= j) || (start[i] == 'R' && i <= j))) {
                i ++; j ++;
            } else return false;

        }
        return i == j;           
    }
};

File: python/algo/202511/algo0777.py
# 滑动窗口与双指针 - 双序列双指针 - 4.1 双指针
class Solution:
    def canTransform(self, start: str, result: str) -> bool:
        # 同 2337
        n = len(start)
        i = j = 0

        while i < n or j < n:
            while i < n and start[i] == 'X':
                i += 1
            while j < n and result[j] == 'X':
                j += 1
            
            if j == n or i == n: break

            if (start[i] == result[j] == 'L' and j <= i ) or start[i] == result[j] == 'R' and j >= i:
                i += 1
                j += 1
            else:
                return False

        return i == j        